File: go/cmd/prompd/namespace.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func handleNamespace() {
	if len(os.Args) < 3 {
		printNamespaceUsage()
		os.Exit(1)
	}

	subcommand := os.Args[2]

	// Handle help flags
	if subcommand == "-h" || subcommand == "--help" {
		printNamespaceUsage()
		return
	}

	switch subcommand {
	case "list":
		handleNamespaceList()
	case "current":
		handleNamespaceCurrent()
	case "use":
		handleNamespaceUse()
	case "create":
		handleNamespaceCreate()
	default:
		fmt.Printf("Unknown namespace subcommand: %s\n", subcommand)
		printNamespaceUsage()
		os.Exit(1)
	}
}

func handleNamespaceList() {
	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("ERROR: Failed to load config: %v\n", err)
		os.Exit(1)
	}

	if len(config.Namespaces) == 0 {
		fmt.Println("No namespaces configured")
		return
	}

	fmt.Println("Available namespaces:")
	for namespace, registry := range config.Namespaces {
		current := ""
		if namespace == config.CurrentNamespace {
			current = " (current)"
		}
		fmt.Printf("  %s → %s%s\n", namespace, registry, current)
	}
}

func handleNamespaceCurrent() {
	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("ERROR: Failed to load config: %v\n", err)
		os.Exit(1)
	}

	if config.CurrentNamespace == "" {
		fmt.Println("No current namespace set")
		return
	}

	registry := config.Namespaces[config.CurrentNamespace]
	fmt.Printf("Current namespace: %s\n", config.CurrentNamespace)
	fmt.Printf("Registry: %s\n", registry)
}

func handleNamespaceUse() {
	if len(os.Args) < 4 {
		fmt.Println("ERROR: namespace use requires a namespace argument")
		fmt.Println("Usage: prompd namespace use @namespace")
		os.Exit(1)
	}

	namespace := os.Args[3]
	if !strings.HasPrefix(namespace, "@") {
		namespace = "@" + namespace
	}

	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("ERROR: Failed to load config: %v\n", err)
		os.Exit(1)
	}

	// Check if namespace exists
	if _, exists := config.Namespaces[namespace]; !exists {
		fmt.Printf("ERROR: Namespace '%s' not found\n", namespace)
		fmt.Println("Use 'prompd namespace list' to see available namespaces")
		os.Exit(1)
	}

	// Update current namespace
	config.CurrentNamespace = namespace
	if err := saveConfig(config); err != nil {
		fmt.Printf("ERROR: Failed to save config: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("✓ Switched to namespace: %s\n", namespace)
}

func handleNamespaceCreate() {
	if len(os.Args) < 4 {
		fmt.Println("ERROR: namespace create requires a namespace argument")
		fmt.Println("Usage: prompd namespace create @namespace [--registry <url>]")
		os.Exit(1)
	}

	namespace := os.Args[3]
	if !strings.HasPrefix(namespace, "@") {
		namespace = "@" + namespace
	}

	// Parse optional registry flag
	registryURL := ""
	for i := 4; i < len(os.Args); i++ {
		if os.Args[i] == "--registry" && i+1 < len(os.Args) {
			registryURL = os.Args[i+1]
			i++
		}
	}

	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("ERROR: Failed to load config: %v\n", err)
		os.Exit(1)
	}

	// Check if namespace already exists
	if _, exists := config.Namespaces[namespace]; exists {
		fmt.Printf("ERROR: Namespace '%s' already exists\n", namespace)
		os.Exit(1)
	}

	// Use default registry if not specified
	if registryURL == "" {
		// Try to get default registry from config
		if defaultReg, ok := config.Registry["default"].(string); ok && defaultReg != "" {
			if registries, ok := config.Registry["registries"].(map[string]interface{}); ok {
				if regInfo, ok := registries[defaultReg].(map[string]interface{}); ok {
					if url, ok := regInfo["url"].(string); ok {
						registryURL = url
					}
				}
			}
		}
		if registryURL == "" {
			registryURL = "https://registry.prompdhub.ai"
		}
	}

	// Add namespace
	if config.Namespaces == nil {
		config.Namespaces = make(map[string]string)
	}
	config.Namespaces[namespace] = registryURL

	// Set as current if it's the first namespace
	if config.CurrentNamespace == "" {
		config.CurrentNamespace = namespace
	}

	if err := saveConfig(config); err != nil {
		fmt.Printf("ERROR: Failed to save config: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("✓ Created namespace: %s\n", namespace)
	fmt.Printf("  Registry: %s\n", registryURL)
}

func printNamespaceUsage() {
	fmt.Println(`Usage: prompd namespace <subcommand> [options]

Manage package namespaces

Subcommands:
  list                      List all configured namespaces
  current                   Show current namespace
  use <namespace>           Switch to a different namespace
  create <namespace>        Create a new namespace

Options:
  --registry <url>          Registry URL (for create command)

Examples:
  prompd namespace list
  prompd namespace current
  prompd namespace use @mycompany
  prompd namespace create @mycompany --registry https://registry.example.com`)
}

File: typescript/src/commands/namespace.ts
import { Command } from 'commander';
import { ConfigManager } from '../lib/config';

export function createNamespaceCommand(): Command {
  const namespaceCommand = new Command('namespace');

  namespaceCommand
    .description('Manage package namespaces');

  // List subcommand
  const listCommand = new Command('list');
  listCommand
    .description('List all configured namespaces')
    .action(async () => {
      const config = new ConfigManager();
      const currentConfig = await config.loadConfig();

      if (!currentConfig.namespaces || Object.keys(currentConfig.namespaces).length === 0) {
        console.log('No namespaces configured');
        return;
      }

      console.log('Available namespaces:');
      for (const [namespace, registry] of Object.entries(currentConfig.namespaces)) {
        const current = namespace === currentConfig.currentNamespace ? ' (current)' : '';
        console.log(`  ${namespace} → ${registry}${current}`);
      }
    });

  // Current subcommand
  const currentCommand = new Command('current');
  currentCommand
    .description('Show current namespace')
    .action(async () => {
      const config = new ConfigManager();
      const currentConfig = await config.loadConfig();

      if (!currentConfig.currentNamespace) {
        console.log('No current namespace set');
        return;
      }

      const registry = currentConfig.namespaces?.[currentConfig.currentNamespace] || 'unknown';
      console.log(`Current namespace: ${currentConfig.currentNamespace}`);
      console.log(`Registry: ${registry}`);
    });

  // Use subcommand
  const useCommand = new Command('use');
  useCommand
    .description('Switch to a different namespace')
    .argument('<namespace>', 'Namespace to use (@namespace format)')
    .action(async (namespace: string) => {
      let ns = namespace;
      if (!ns.startsWith('@')) {
        ns = '@' + ns;
      }

      const config = new ConfigManager();
      const currentConfig = await config.loadConfig();

      if (!currentConfig.namespaces || !currentConfig.namespaces[ns]) {
        console.error(`ERROR: Namespace '${ns}' not found`);
        console.log('Use \'prompd namespace list\' to see available namespaces');
        process.exit(1);
      }

      currentConfig.currentNamespace = ns;
      await config.saveConfig(currentConfig);

      console.log(`✓ Switched to namespace: ${ns}`);
    });

  // Create subcommand
  const createCommand = new Command('create');
  createCommand
    .description('Create a new namespace')
    .argument('<namespace>', 'Namespace to create (@namespace format)')
    .option('-r, --registry <url>', 'Registry URL')
    .action(async (namespace: string, options: { registry?: string }) => {
      let ns = namespace;
      if (!ns.startsWith('@')) {
        ns = '@' + ns;
      }

      const config = new ConfigManager();
      const currentConfig = await config.loadConfig();

      // Check if namespace already exists
      if (currentConfig.namespaces && currentConfig.namespaces[ns]) {
        console.error(`ERROR: Namespace '${ns}' already exists`);
        process.exit(1);
      }

      // Determine registry URL
      let registryURL = options.registry;
      if (!registryURL) {
        if (currentConfig.registry?.default) {
          const defaultRegistry = currentConfig.registry.registries?.[currentConfig.registry.default];
          if (defaultRegistry) {
            registryURL = defaultRegistry.url;
          }
        }
        if (!registryURL) {
          registryURL = 'https://registry.prompdhub.ai';
        }
      }

      // Add namespace
      if (!currentConfig.namespaces) {
        currentConfig.namespaces = {};
      }
      currentConfig.namespaces[ns] = registryURL;

      // Set as current if it's the first namespace
      if (!currentConfig.currentNamespace) {
        currentConfig.currentNamespace = ns;
      }

      await config.saveConfig(currentConfig);

      console.log(`✓ Created namespace: ${ns}`);
      console.log(`  Registry: ${registryURL}`);
    });

  namespaceCommand.addCommand(listCommand);
  namespaceCommand.addCommand(currentCommand);
  namespaceCommand.addCommand(useCommand);
  namespaceCommand.addCommand(createCommand);

  return namespaceCommand;
}
